samples/api/timestamp_queries/README.md

```diff
@@ -0,0 +1,184 @@
+
+```cpp
+vkCmdWriteTimestamp(draw_cmd_buffers[i], VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT, query_pool_timestamps, 0);
+// Do some work
+for (int i = 0; i < draw_call_count; i++) {
+	vkCmdDraw(...);
+}
+vkCmdWriteTimestamp(draw_cmd_buffers[i], VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT, query_pool_timestamps, 1);
+```
+
+To measure GPU times for the draw calls(s) we first tell the GPU to write a timestamp at the `VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT` pipeline stage. This is not a real pipeline stage (as in e.g. the vertex or fragment stages) but a special constant that tells the GPU to wait for all commands currently in flight to finish. This ensures, that we get a timestamp right before starting on the draw calls we want to measure. This timestamp will then be the base for calculating a delta time.
+
+The second timestamp is written at the `VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT` pipeline stage. Once again this is not a real pipeline stage, but it tells the GPU to wait for all work to be finished. 
+
+## Getting the results
+
+Reading back the results can be done in two ways:
+
+- Copy the results into a `VkBuffer` inside the command buffer using `vkCmdCopyQueryPoolResults`
+- Get the results after the command buffer has finished executing using `vkGetQueryPoolResults`
+
+For our sample we'll use option two (see `get_time_stamp_results`):
+
+```cpp
+vkQueueSubmit();
+...
+// The number of timestamps changes if the bloom pass is disabled
+uint32_t count = bloom ? time_stamps.size() : time_stamps.size() - 2;
+
+vkGetQueryPoolResults(
+	device->get_handle(),
+	query_pool_timestamps,
+	0,
+	count,
+	time_stamps.size() * sizeof(uint64_t),
+	time_stamps.data(),
+	sizeof(uint64_t),
+	VK_QUERY_RESULT_64_BIT | VK_QUERY_RESULT_WAIT_BIT);
+```
+
+Most arguments are straightforward, e.g. where the data will be copied to (the `time_stamps` vector). The important part here are the `VK_QUERY_RESULT_` flags used here.
+
+`VK_QUERY_RESULT_64_BIT` will tell the api that we want to get the results as 64 bit values. Without this flag, we would only get 32 bit values. And since timestamp queries can operate in nanoseconds, only using 32 bits could result into an overflow. E.g. if your device has a `timestampPeriod` of 1, so that one increment in the result maps to exactly one nanosecond, with 32 bit precision you'd run into such an overflow after only about 0.43 seconds.
+
+The `VK_QUERY_RESULT_WAIT_BIT` bit then tells the api to wait for all results to be available. So when using this flag the values written to our `time_stamps` vector is guaranteed to be available after calling `vkGetQueryPoolResults`. This is fine for our use-case where we want to immediately access the results, but may introduce unnecessary stalls in other scenarios. An alternative is using the `VK_QUERY_RESULT_WITH_AVAILABILITY_BIT`, which will let you poll the availability of the results and defer writing new timestamps until the results are available.
```

Comment:
Adding an alternative code path to the sample probably won't work. Using the availability flag you'd need to not write timestamps in the command buffer until the results are available. Since our samples pre-built the command buffers, this isn't easy to implement.

I'll elaborate a bit on this in the readme.

Reply:
I added some more details on that flag along with a code example, please see https://github.com/KhronosGroup/Vulkan-Samples/blob/dcfb2f6cb446ad38375d3e67e392d8074c2340f7/samples/api/timestamp_queries/README.md#getting-the-results